### PR TITLE
Prompt formatter API and canary transcribe tensor input support

### PR DESCRIPTION
**Pull request by pzelasko**
# What does this PR do ?

Generic prompt formatter for text modality with several out-of-the-box prompt format definitions. See the class documentation for more details.

Also, enables support for tensor/array inputs in Canary. Example snippet:

```python
import os
from tempfile import NamedTemporaryFile

import lhotse

from nemo.collections.asr.models.aed_multitask_models import EncDecMultiTaskModel


def main():
    model = EncDecMultiTaskModel.from_pretrained("nvidia/canary-1b")

    path = ...
    rec = lhotse.Recording.from_file(path)
    audio = rec.load_audio()[0]

    TARGET_LANG = "es"

    # Array list input, legacy API for Canary-1B
    results = model.transcribe(
        audio=[audio],
        batch_size=1,
        num_workers=0,
        source_lang="en",
        target_lang=TARGET_LANG,
        task="asr",
        pnc="yes",
    )

    print(results)

    # Array list input, explicit single-turn prompt   
    results = model.transcribe(
        audio=[audio],
        batch_size=1,
        num_workers=0,
        role="user",
        slots={
            "source_lang": "en",
            "target_lang": TARGET_LANG,
            "task": "asr",
            "pnc": "yes",
        },
    )

    print(results)

    # Array list input, explicit multi-turn prompt   
    results = model.transcribe(
        audio=[audio],
        batch_size=1,
        num_workers=0,
        turns=[{
            "role": "user",
            "slots": {
                "source_lang": "en",
                "target_lang": TARGET_LANG,
                "task": "asr",
                "pnc": "yes",
            },
        }],
    )

    print(results)

    # Audio path input, explicit single-turn prompt   
    results = model.transcribe(
        audio=path,
        batch_size=1,
        num_workers=0,
        role="user",
        slots={
            "source_lang": "en",
            "target_lang": TARGET_LANG,
            "task": "asr",
            "pnc": "yes",
        },
    )

    print(results)

    # Legacy JSON manifest with slot values API for Canary-1B 
    with NamedTemporaryFile("w", suffix=".json") as f:
        lhotse.serialization.save_to_jsonl(
            [
                {
                    "audio_filepath": path,
                    "text": "irrelevant",
                    "duration": rec.duration,
                    "task": "asr",
                    "pnc": "yes",
                    "source_lang": "en",
                    "target_lang": TARGET_LANG,
                }
            ],
            f.name,
        )
        f.flush()
        os.fsync(f.fileno())

        results = model.transcribe(
            audio=f.name,
            batch_size=1,
            num_workers=0,
        )

        print(results)


if __name__ == "__main__":
    main()

```

We can also now provide these values dynamically to `transcribe_speech.py` from CLI. Example:

```bash
# Legacy Canary-1B format
python ~/code/NeMo/examples/asr/transcribe_speech.py \
  audio_dir=wavs \
  output_filename=out.json \
  batch_size=1 \
  pretrained_name=nvidia/canary-1b \
  +prompt.source_lang=en \
  +prompt.target_lang=es \
  +prompt.task=asr \
  +prompt.pnc=yes

# Explicit single-turn format
python ~/code/NeMo/examples/asr/transcribe_speech.py \
  audio_dir=wavs \
  output_filename=out.json \
  batch_size=1 \
  pretrained_name=nvidia/canary-1b \
  +prompt.role=user \
  +prompt.slots.source_lang=en \
  +prompt.slots.target_lang=es \
  +prompt.slots.task=asr \
  +prompt.slots.pnc=yes

# Explicit multi-turn format
python ~/code/NeMo/examples/asr/transcribe_speech.py \
  audio_dir=wavs \
  output_filename=out.json \
  batch_size=1 \
  pretrained_name=nvidia/canary-1b \
  +prompt.turns='[{role:user,slots:{source_lang:en,target_lang:es,task:asr,pnc:yes}}]'
```

**Collection**: ASR

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Review by titu1994 — COMMENTED** [review]
Initial comments

File: nemo/collections/common/prompts/canary.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
+from nemo.collections.common.prompts.formatter import PromptFormatter
```
[the remark this answers is not in the document]
Comment (pzelasko) [reply nemo/collections/common/prompts/canary.py]: I think common should be fine as we'll definitely use this multimodal and maybe even get somebody from nlp to try it as well. Note we don't import anything from ASR as the only dependency is tokenizer (also in common).

File: nemo/collections/common/prompts/canary.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        INFERENCE_ROLE: {
+            "template": f"|TEXT|{CANARY_EOS}",
+            "slots": {
+                "|TEXT|": str,
+            },
+        },
+    }
+
+
+def map_manifest_values_to_special_tokens(slot_values: dict[str, str]) -> dict[str, str]:
```
[the remark this answers is not in the document]
Comment (pzelasko) [reply nemo/collections/common/prompts/canary.py:45]: done

File: nemo/collections/asr/parts/utils/streaming_utils.py [hunk 3]
```diff
@@ -1588,15 +1611,23 @@ def get_input_tokens(self, sample: dict):
...
-            tokens = canary_prompt(
-                tokenizer=self.asr_model.tokenizer,
-                text=None,
-                language=None,
-                source_language=sample['source_lang'],
-                target_language=sample['target_lang'],
-                taskname=sample['taskname'],
-                pnc=sample['pnc'],
-            )
+            formatter = CanaryPromptFormatter(self.asr_model.tokenizer)
```
[the remark this answers is not in the document]
Comment (pzelasko) [reply nemo/collections/asr/parts/utils/streaming_utils.py:1614]: good point will update

File: nemo/collections/common/prompts/formatter.py [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+
+    Another example that shows how to use the same method to generate prompts for inference::
+
+
+        >>> formatter = PromptFormatter(tokenizer)
+        ... encoded_for_training = formatter.encode_dialog(
+        ...     turns=[
+        ...         {"role": "user", "slots": {"message": "What time is it?"}},
+        ...         {"role": "assistant", "slots": {"message": "Ten o'clock."}},
+        ...         {"role": "user", "slots": {"message": "PM or AM?"}},
```
Comment (titu1994) [inline nemo/collections/common/prompts/formatter.py:88]: Excellent docs !

File: nemo/collections/common/prompts/formatter.py [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+    def get_template(cls, role: str) -> str:
+        return cls.TEMPLATE[role]["template"]
+
+    def get_default_dialog_slots(self) -> list[dict]:
+        """
+        Returns a list of dialog turns that can be used as a skeleton to fill with actual slot values.
+        If ``PromptFormatter`` was initialized with ``defaults`` argument, this method will return the
+        defaults. Otherwise, every slot is pre-filled with ``None``.
+        """
+        if self._defaults is not None:
```
Comment (titu1994) [inline nemo/collections/common/prompts/formatter.py:172]: Hmm I was thinking of not having to pass fully defined default. So basically in else last, for each slot in each role, see if that role exists in defaults and if slot exists in default[role] and copy it into result default slots dict, else set None

File: tests/collections/common/prompt_formatters/conftest.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+So, it might be a good idea to create a new tokenizer instead when adding new prompt formats.
+"""
+
+
+@pytest.fixture(scope="session")
+def bpe_tokenizer(tmp_path_factory):
+    tmpdir = tmp_path_factory.mktemp("bpe_tokenizer")
+    text_path = tmpdir / "text.txt"
+    text_path.write_text(TOKENIZER_TRAIN_TEXT)
+    create_spt_model(str(text_path), vocab_size=512, sample_size=-1, do_lower_case=False, output_dir=str(tmpdir))
```
Comment (titu1994) [inline tests/collections/common/prompt_formatters/conftest.py:38]: Doesn't this take a long time ?

**Review by titu1994 — COMMENTED** [review]
Overall, really impressive pr! This should work for a lot of future projects in ASR and speechlm

File: tests/collections/common/prompt_formatters/conftest.py (hunk 6 above)
Comment (pzelasko) [reply tests/collections/common/prompt_formatters/conftest.py:38]: it's quick, takes 1-2 seconds, but I'm running into the need of setting tokenizer fixtures repeatedly and I think it would be nice to create a tokenizer fixture builder where you just fix its vocabulary, but otherwise it works as a regular BPE tokenizer. unless sth like that already exists in NeMo?

File: examples/asr/transcribe_speech.py [hunk 7]
```diff
@@ -169,6 +169,8 @@ class TranscriptionConfig:
 
     # Decoding strategy for AED models
     multitask_decoding: MultiTaskDecodingConfig = MultiTaskDecodingConfig()
+    # Prompt slots for prompted models, e.g. for Canary-1B: +prompt.user.source_lang=en
+    prompt_slots: Optional[dict[str, dict[str, str]]] = None
```
[the remark this answers is not in the document]
Comment (pzelasko) [reply examples/asr/transcribe_speech.py:173]: good catch

File: nemo/collections/asr/data/audio_to_text_lhotse_prompted.py [hunk 8]
```diff
@@ -141,107 +143,41 @@ def canary(cuts: CutSet, tokenizer: TokenizerWrapper, inference: bool = False) -
...
+            ]
+        )
+        tokens.append(encoded["input_ids"])
+        prompts.append(encoded["context_ids"])
 
-    # eos (only in training)
-    if tokens is not None:
-        prompted_tokens.append(tokenizer.eos_id)
-    return prompted_tokens
+    return tokens, prompts
```
[the remark this answers is not in the document]
Comment (pzelasko) [reply nemo/collections/asr/data/audio_to_text_lhotse_prompted.py:180]: prompts for inference (as opposed to prompts for training), this is how we got canary transcribe to work so far. note that `.transcribe(manifest_path)` and also `.transcribe(audio_path)` still go through the prompted.bla.bla.dataset route. could be a good refactor follow up opportunity. 

for now, I'll rename these fields to make this clearer

File: nemo/collections/asr/data/audio_to_text.py [hunk 9]
```diff
@@ -75,7 +75,9 @@ def _speech_collate_fn(batch, pad_id):
     has_audio = audio_lengths[0] is not None
     if has_audio:
         max_audio_len = max(audio_lengths).item()
-    max_tokens_len = max(tokens_lengths).item()
+    has_tokens = tokens_lengths[0] is not None
```
Comment (stevehuang52) [inline nemo/collections/asr/data/audio_to_text.py:78]: what's the motivation for this change?

File: nemo/collections/asr/models/aed_multitask_models.py [hunk 10]
```diff
@@ -134,6 +131,12 @@ def __init__(self, cfg: DictConfig, trainer: Trainer = None):
 
         super().__init__(cfg=cfg, trainer=trainer)
 
+        prompt_cls = PromptFormatter.resolve(self.prompt_format)
+        self.prompt = prompt_cls(
```
[the remark this answers is not in the document]
Comment (pzelasko) [reply nemo/collections/asr/models/aed_multitask_models.py:135]: can you elaborate? not sure what to do here

File: nemo/collections/asr/models/aed_multitask_models.py [hunk 11]
```diff
@@ -718,6 +716,8 @@ def _transcribe_on_begin(self, audio, trcfg: MultiTaskTranscriptionConfig):
         # Switch model to evaluation mode
         self.transf_decoder.freeze()
 
+        self.prompt_formatter = PromptFormatter.resolve(self.prompt_format)(self.tokenizer)
```
[the remark this answers is not in the document]
Comment (pzelasko) [reply nemo/collections/asr/models/aed_multitask_models.py:719]: good catch. it's a left-over from earlier version, every method should be using `self.prompt` instead which is initialized in the model's constructor

File: nemo/collections/asr/models/aed_multitask_models.py [hunk 12]
```diff
@@ -913,28 +933,18 @@ def _may_be_make_dict_and_fix_paths(self, json_items, manifest_path, trcfg: Mult
...
-                    entry['taskname'] = 'asr' if trcfg.task is None else trcfg.task
-                if 'target_lang' not in entry:
-                    entry['target_lang'] = 'en' if trcfg.target_lang is None else trcfg.target_lang
-                if 'pnc' not in entry:
-                    entry['pnc'] = 'yes' if trcfg.pnc is None else 'yes' if trcfg.pnc else 'no'
                 if trcfg.text_field not in entry:
                     entry[trcfg.text_field] = 'nothing'
             else:
                 raise ValueError(f"Expected str or dict, got {type(item)}")
+            for k, dv in (("source_lang", "en"), ("target_lang", "en"), ("taskname", "asr"), ("pnc", "yes")):
```
[the remark this answers is not in the document]
Comment (pzelasko) [reply nemo/collections/asr/models/aed_multitask_models.py:945]: yes but note that it wouldn't work with the released canary ckpt because it doesn't have the prompt defaults in config; I can add what you're proposing but we probably need to keep this part anyway when self.prompt_format=="canary"

File: nemo/collections/common/prompts/canary.py [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+        # but we are not using it here anymore.
+        # This maps things such as '|task|: "asr"' to '|TASK|: "<|transcribe|>"'.
+        slot_values = map_manifest_values_to_special_tokens(slot_values)
+        return super().encode_turn(
+            prompt_template=prompt_template, expected_slots=expected_slots, slot_values=slot_values
+        )
+
+
+def map_manifest_values_to_special_tokens(slot_values: dict[str, str]) -> dict[str, str]:
+    slot_values = slot_values.copy()
```
[the remark this answers is not in the document]
Comment (pzelasko) [reply nemo/collections/common/prompts/canary.py:46]: no, all values are references to the values in the original dict. shallow copy only prevents the original dict from being modified when you assign values to keys

File: nemo/collections/common/prompts/formatter.py (hunk 5 above)
Comment (pzelasko) [reply nemo/collections/common/prompts/formatter.py:172]: good point

File: nemo/collections/asr/data/audio_to_text.py (hunk 9 above)
Comment (pzelasko) [reply nemo/collections/asr/data/audio_to_text.py:78]: yes, it now supports in transcribe() tensor dataset which may return None for token/lens instead of dummy tensor

File: nemo/collections/asr/data/audio_to_text_lhotse_prompted.py [hunk 14]
```diff
@@ -132,116 +135,50 @@ def canary(cuts: CutSet, tokenizer: TokenizerWrapper, inference: bool = False) -
...
     ), "To use 'canary' prompt format, you must use the CanaryTokenizer."
-    tokenizer = tokenizer._tokenizer
+    formatter = CanaryPromptFormatter(tokenizer._tokenizer)
 
-    tokens, prompts = [], []
+    prompts_with_answers, prompts = [], []
     for cut in cuts:
         if isinstance(cut, MixedCut):
             cut = cut._first_non_padding_cut
         assert isinstance(cut, MonoCut), "Expected MonoCut."
```
Comment (stevehuang52) [inline nemo/collections/asr/data/audio_to_text_lhotse_prompted.py:144]: Better change to raising TypeError and saying something like "expected input audio to have single channel", since users might not know what "MoNoCut" means

File: nemo/collections/asr/data/audio_to_text_lhotse_prompted.py [hunk 15]
```diff
@@ -105,7 +106,9 @@ def get_prompt_format_fn(name: str) -> Callable[[CutSet, TokenizerWrapper, bool]
```
Comment (stevehuang52) [inline nemo/collections/asr/data/audio_to_text_lhotse_prompted.py:106]: Shall we consider moving to `PROMPT_FORMAT_FNS`, `registered_prompt_format_fn` and `get_prompt_format_fn` to ` nemo.collections.common.prompts`? Seems that these are more general than the audio to text datasets. Also, should it be `registered_prompt_format_fn` or `register_prompt_format_fn`?

File: nemo/collections/common/prompts/formatter.py [hunk 16]
```diff
@@ -0,0 +1,294 @@
...
+    #       * keys of slots are the names of formattable slots in the prompt template
+    #       * values of slots are :class:`Modality` objects that can be used to check
+    #           whether a specific value conforms to a given modality requirements
+    #           (e.g., Modality.Text may expect string objects).
+    # Template is intended to be defined by the child classes.
+    TEMPLATE = None
+
+    # Turns under this role indicate responses by the model; if the last turn in
+    # PromptFormatter.encode_dialog() ends with this role, it indicates a training example.
+    INFERENCE_ROLE = None
```
Comment (stevehuang52) [inline nemo/collections/common/prompts/formatter.py:114]: I wonder if we can use a better name than `INFERENCE_ROLE`, maybe `MODEL_ROLE`, `SYSTEM_ROLE` or something else? Personally I feel `INFERENCE_ROLE` a bit strange

File: nemo/collections/common/prompts/formatter.py [hunk 17]
```diff
@@ -0,0 +1,294 @@
...
+        ...     ]
+        ... )
+
+    """
+
+    # Used to support AggregateTokenizer; this key selects the right sub-tokenizer for each turn.
+    PROMPT_LANGUAGE_SLOT = "prompt_language"
+
+    # Sub-classes will be registered under this name, to be used via PromptFormatter.resolve(name).
+    REGISTER_NAME = None
```
Comment (stevehuang52) [inline nemo/collections/common/prompts/formatter.py:98]: Maybe `REGISTERED_NAME` , or `REGISTRY`?

File: nemo/collections/common/prompts/formatter.py [hunk 18]
```diff
@@ -0,0 +1,294 @@
...
+        If ``PromptFormatter`` was initialized with ``defaults`` argument, this method will return the
+        defaults. Otherwise, every slot is pre-filled with ``None``.
+        """
+        if self._defaults is not None:
+            return self._defaults.copy()
+        else:
+            return [
+                {"role": role, "slots": {slot: None for slot in self.get_slots(role)}}
+                for role in self.get_roles()
+                if role != self.INFERENCE_ROLE
```
Comment (stevehuang52) [inline nemo/collections/common/prompts/formatter.py:178]: If I understand properly, this function provides an example of what the input dict should should like, which includes both the user's and the system's roles, right? If so, why do we exclude the INFERENCE_ROLE?

File: nemo/collections/common/prompts/formatter.py [hunk 19]
```diff
@@ -0,0 +1,294 @@
...
+        for slot in expected_slots:
+            # For the final substitution of 'slot' in the template we have to mangle it to '|slot|' anyway,
+            # but 'slot' form enables to use valid python identifiers as **kwargs
+            # for passing slots around in user functions.
+            value = slot_values.get(slot)
+            assert value is not None, f"Missing required {slot=} in {slot_values=} for {prompt_template=}"
+            prompt = prompt.replace(_mangled(slot), value)
+        return self._apply_tokenizer(prompt, lang=slot_values.get(self.PROMPT_LANGUAGE_SLOT))
+
+    def encode_dialog(self, turns: list[dict]) -> dict[str, torch.Tensor]:
```
Comment (krishnacpuvvada) [inline nemo/collections/common/prompts/formatter.py:194]: If i understand correctly, this PR is for encoder-decoder models like canary/bestow, where all the (multi-turn) dialogue should be in text. 
can we think little bit about supporting audio modality also in slot values. (may be we should keep audio slots untokenized and replace it with "audio features" later. one way is for prompt formatter to return something like list of lists)

Comment (zhehuaichen) [reply nemo/collections/common/prompts/formatter.py:194]: +1 to this and Piotr told me he is planning on this as v2. maybe we can resume the discussion at the time

**Review by zhehuaichen — COMMENTED** [review]
Nice work! LGTM

File: nemo/collections/common/prompts/formatter.py (hunk 19 above)
Comment (krishnacpuvvada) [reply nemo/collections/common/prompts/formatter.py:194]: sg. 
if possible, lets try to put the skeleton in place, e.g. if slot value needs to re-defined as (value, modality) tuple, return needs to be list of lists/tuples etc.

File: nemo/collections/asr/data/audio_to_text_lhotse_prompted.py (hunk 15 above)
Comment (pzelasko) [reply nemo/collections/asr/data/audio_to_text_lhotse_prompted.py:106]: TBH I think maybe we can get rid of those entirely with just a bit more refactoring but I am reluctant to do it in this PR as the scope is already wide.

File: nemo/collections/common/prompts/formatter.py (hunk 16 above)
Comment (pzelasko) [reply nemo/collections/common/prompts/formatter.py:114]: system is ambiguous because of "system prompt" in llama which is not the system's response, so I discarded that option. model - maybe? I find inference the least ambiguous. WDYT?

File: nemo/collections/common/prompts/formatter.py (hunk 18 above)
Comment (pzelasko) [reply nemo/collections/common/prompts/formatter.py:178]: it doesn't include the "assistant's response" (not to mix up with "system prompt"). it provides the user an easy way of creating a prompt for model inference. It doesn't make sense to have a default prompt for training as you need to fill the slots according to each training example

File: nemo/collections/common/prompts/formatter.py (hunk 19 above)
Comment (krishnacpuvvada) [reply nemo/collections/common/prompts/formatter.py:194]: sg. 
Agreed. any audio encoder (especially our 600M ones) has to be run on GPU.

File: nemo/collections/common/prompts/__init__.py
Comment (krishnacpuvvada) [inline nemo/collections/common/prompts/__init__.py:1]: a thought.. can we also a sample.py/simple.py template with the simplest possible template and add few comments about which routines need to be defined. (This is mainly coming from - if  a user wants to create their own custom template; I know there are plenty of examples already.. )

File: nemo/collections/common/prompts/formatter.py (hunk 16 above)
Comment (stevehuang52) [reply nemo/collections/common/prompts/formatter.py:114]: hmmm.....personally I feel INFERENCE is not very straightforward, since both the model's output and user's input are taken into inference process, and I didn't understand what INFERENCE_ROLE mean until I read the documents. maybe @krishnacpuvvada @titu1994 can provide more suggestions?

Comment (pzelasko) [reply nemo/collections/common/prompts/formatter.py:114]: maybe ASSISTANT_ROLE or RESPONSE_ROLE...

File: nemo/collections/common/prompts/formatter.py (hunk 5 above)
Comment (titu1994) [reply nemo/collections/common/prompts/formatter.py:172]: It's a good user experience not to demand fully specified args if they can be defaulted. For example we don't always provide be in transcribe, or pnc

File: tests/collections/common/prompt_formatters/conftest.py (hunk 6 above)
Comment (titu1994) [reply tests/collections/common/prompt_formatters/conftest.py:38]: Doesn't exist I think

**Review by titu1994 — COMMENTED** [review]
It looks really good now, minor comments from me, lets address the rest and merge

File: nemo/collections/asr/models/aed_multitask_models.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (pzelasko) [reply nemo/collections/asr/models/aed_multitask_models.py:959]: > For backward compat, you can add to the init of the model - cfg.get("prompt", "canary")?

This is not needed, because we are leveraging the existing `cfg.prompt_format` key in the config to initialize the prompt formatter.

File: nemo/collections/common/prompts/__init__.py
Comment (pzelasko) [reply nemo/collections/common/prompts/__init__.py:1]: OK

File: nemo/collections/common/prompts/formatter.py (hunk 17 above)
Comment (pzelasko) [reply nemo/collections/common/prompts/formatter.py:98]: I think I'll go with just `NAME` for simplicity

File: nemo/collections/common/prompts/formatter.py (hunk 16 above)
Comment (pzelasko) [reply nemo/collections/common/prompts/formatter.py:114]: yeah, but these are terminologies for specific instances of LLMs such as llama. I am still hesitant about any of these names. Let me try OUTPUT_ROLE as that identifies which one is the output of whatever we call the model/system/assistant.

File: nemo/collections/common/prompts/formatter.py (hunk 5 above)
Comment (pzelasko) [reply nemo/collections/common/prompts/formatter.py:172]: OK

File: nemo/collections/common/prompts/formatter.py (hunk 16 above)
Comment (titu1994) [reply nemo/collections/common/prompts/formatter.py]: Seems reasonable

**Review by titu1994 — APPROVED** [review]
Amazing work !

**Review by stevehuang52 — APPROVED** [review]
Great work~! LGTM
